File: Current/Product/Production/Common/Common.Tasks/SwitchTask.vb
Option Explicit On
Option Strict On

Imports System.Xml
Imports System.Reflection

Imports NAnt.Core
Imports NAnt.Core.Types
Imports NAnt.Core.Tasks
Imports NAnt.Core.Attributes

<TaskName("switch")> _
Public Class SwitchTask
    Inherits Task

    Private _Cases As CaseContainerCollection
    Private _LeftValue As String
    Private _Else As TaskContainer

    <TaskAttribute("value", REquired:=True)> _
    Public Property LeftValue() As String
        Get
            Return _LeftValue
        End Get
        Set(ByVal value As String)
            _LeftValue = value
        End Set
    End Property

    <BuildElementArray("case", Required:=False)> _
        Public Property Cases() As CaseContainerCollection
        Get
            If _Cases Is Nothing Then
                _Cases = New CaseContainerCollection()
            End If
            Return _Cases
        End Get
        Set(ByVal value As CaseContainerCollection)
            _Cases = value
        End Set
    End Property

    <BuildElement("default", Required:=False)> _
        Public Property [Else]() As TaskContainer
        Get
            Return _Else
        End Get
        Set(ByVal value As TaskContainer)
            _Else = value
        End Set
    End Property

    Protected Overrides Sub ExecuteTask()
        For Each [Case] As CaseElement In Me.Cases
            If Me.LeftValue = [Case].RightValue Then
                [Case].Execute()
                If [Case].Break Then
                    Exit Sub
                End If
            End If
        Next
        Me.Else.Execute()
    End Sub

End Class

File: tags/1.0.0.19/Product/Production/Nant/CIFactory.NAnt.Tasks/Tasks/SwitchTask.cs
using System;
using NAnt.Core;
using NAnt.Core.Attributes;
using CIFactory.NAnt.Types;

namespace CIFactory.NAnt.Tasks
{
    [TaskName("switch")]
    public class SwitchTask : Task
    {
        #region Fields

        private CaseContainerCollection _Cases;

        private TaskContainer _Else;

        private string _LeftValue;

        #endregion

        #region Properties

        [BuildElementArray("case", Required = false)]
        public CaseContainerCollection Cases
        {
            get
            {
                if (_Cases == null)
                {
                    _Cases = new CaseContainerCollection();
                }
                return _Cases;
            }
            set { _Cases = value; }
        }

        [BuildElement("default", Required = false)]
        public TaskContainer Else
        {
            get { return _Else; }
            set { _Else = value; }
        }

        [TaskAttribute("value", Required = true)]
        public string LeftValue
        {
            get { return _LeftValue; }
            set { _LeftValue = value; }
        }

        #endregion

        #region Protected Methods

        protected override void ExecuteTask()
        {
            foreach (CaseElement Case in this.Cases)
            {
                if (this.LeftValue == Case.RightValue)
                {
                    Case.Execute();
                    if (Case.Break)
                    {
                        return; // TODO: might not be correct. Was : Exit Sub
                    }
                }
            }
            this.Else.Execute();
        }

        #endregion

    }

}